agents/dqn.py

```diff
@@ -1,137 +1,132 @@
-from typing import Dict, Any
+from copy import deepcopy
+from typing import Dict, Any, Callable, Optional
 
 import torch
-from torch import Tensor
+from torch import Tensor, nn
 from torch.distributions import Categorical
 
 from agents import TorchAgent
+from ml import BaseOptimizer
 from ml.dqn_encoders import BaseQNetwork
 from ml.encoders import BaseEncoder
-from ml.utils import TensorWithMask
 from utils.bag_trajectory import BaseBagTrajectoryMemory
 
 
-def _with_random_research(
-        argmax_next_neighbor: Tensor,
-        neighbor: TensorWithMask,
-        prob: float
-):
-    probs = torch.stack([
-        torch.full(argmax_next_neighbor.shape, prob),
-        torch.full(argmax_next_neighbor.shape, 1 - prob)
-    ], dim=-1)
-    choice_idx = Categorical(probs=probs).sample()
-    choices = torch.stack([
-        argmax_next_neighbor,
-        neighbor
-    ], dim=-1)
-    return torch.squeeze(torch.gather(choices, dim=-1, index=choice_idx))
-
-
 class DQNAgent(TorchAgent, config_name='dqn'):
     def __init__(
             self,
             current_node_idx_prefix: str,
             destination_node_idx_prefix: str,
             neighbors_node_ids_prefix: str,
+            output_prefix: str,
+            bag_ids_prefix: str,
             q_network: BaseQNetwork,
             discount_factor: float = 0.99,
             research_prob: float = 0.1,
             bag_trajectory_memory: BaseBagTrajectoryMemory = None,
-            node_idx: int = None,
-            sample_size: int = 100,
+            trajectory_sample_size: int = 30,
+            optimizer_factory: Callable[[nn.Module], BaseOptimizer] = None
     ):
         assert 0 < discount_factor < 1, 'Incorrect `discount_factor` choice'
         assert 0 < research_prob < 1, 'Incorrect `discount_factor` choice'
         super().__init__()
 
+        self._node_id = -1
```

Comment:
Better to do it None. But even more better would be to make something like `self._node_id = node_id` where `node_id` is an argument in the `__init__` which is equal `None` by default). In such cases, we can use `None` for pre-train or centralized scenarios while giving precise ID in train and similar scenarios.

Reply:
Let's make it None